lib/src/Uncertainty/Algorithm/WeightedExperiments/SmolyakExperiment.cxx

```diff
@@ -0,0 +1,348 @@
+//                                               -*- C++ -*-
+/**
+ *  @brief Abstract top-level view of an Smolyak experiment
+ *
+ *  Copyright 2005-2023 Airbus-EDF-IMACS-ONERA-Phimeca
+ *
+ *  This library is free software: you can redistribute it and/or modify
+ *  it under the terms of the GNU Lesser General Public License as published by
+ *  the Free Software Foundation, either version 3 of the License, or
+ *  (at your option) any later version.
+ *
+ *  This library is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU Lesser General Public License for more details.
+ *
+ *  You should have received a copy of the GNU Lesser General Public License
+ *  along with this library.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+#include "openturns/SmolyakExperiment.hxx"
+#include "openturns/TensorProductExperiment.hxx"
+#include "openturns/PersistentObjectFactory.hxx"
+#include "openturns/LinearEnumerateFunction.hxx"
+#include "openturns/SpecFunc.hxx"
+#include "openturns/BlockIndependentDistribution.hxx"
+
+BEGIN_NAMESPACE_OPENTURNS
+
+CLASSNAMEINIT(SmolyakExperiment)
+
+/* Default constructor */
+SmolyakExperiment::SmolyakExperiment()
+  : WeightedExperimentImplementation()
+{
+  // Nothing to do
+}
+
+/* Constructor with parameters */
+SmolyakExperiment::SmolyakExperiment(const WeightedExperimentCollection & collection,
+                                     const UnsignedInteger level
+                                    )
+  : WeightedExperimentImplementation()
+  , collection_(collection)
+  , level_(level)
+{
+  if (level == 0) throw InvalidArgumentException(HERE) << "Error: the level is zero";
+  const UnsignedInteger numberOfMarginalExperiments = collection_.getSize();
+  BlockIndependentDistribution::DistributionCollection distributionCollection(numberOfMarginalExperiments);
+  for (UnsignedInteger i = 0; i < numberOfMarginalExperiments; ++ i)
+  {
+    distributionCollection[i] = collection_[i].getDistribution();
+    const UnsignedInteger marginalDimension = distributionCollection[i].getDimension();
+    if (marginalDimension != 1) throw InvalidArgumentException(HERE) << "Error: the marginal with index " << i << " has dimension " << marginalDimension << " which is different from 1";
+    if (collection_[i].isRandom()) throw InvalidArgumentException(HERE) << "Error: the marginal with index " << i << " is random.";
+  }
+  const BlockIndependentDistribution distribution(distributionCollection);
```

Comment:
Why do we need to rely on BlockIndependentDistribution as we only accept 1D marginals? A ComposedDistribution would be more adapted.